// knip/dist/util/glob.js
import fg from 'fast-glob';
import { GLOBAL_IGNORE_PATTERNS } from '../constants.js';
import { timerify } from './Performance.js';
import { compact } from './array.js';
import { globby } from './glob-core.js';
import { join, relative } from './path.js';
const prependDirToPatterns = (cwd, dir, patterns) => {
    const relativePath = relative(cwd, dir);
    const prepend = (pattern) => prependDirToPattern(relativePath, pattern);
    return compact([patterns].flat().map(prepend).map(removeProductionSuffix)).sort(negatedLast);
};
const removeProductionSuffix = (pattern) => pattern.replace(/!$/, '');
const negatedLast = (pattern) => (pattern.startsWith('!') ? 1 : -1);
export const prependDirToPattern = (dir, pattern) => {
    if (pattern.startsWith('!'))
        return `!${join(dir, pattern.slice(1))}`;
    return join(dir, pattern);
};
export const negate = (pattern) => pattern.replace(/^!?/, '!');
export const hasProductionSuffix = (pattern) => pattern.endsWith('!');
export const hasNoProductionSuffix = (pattern) => !pattern.endsWith('!');
const glob = async ({ cwd, dir = cwd, patterns, gitignore = true }) => {
    if (patterns.length === 0)
        return [];
    const globPatterns = prependDirToPatterns(cwd, dir, patterns);
    if (globPatterns[0].startsWith('!'))
        return [];
    return globby(globPatterns, {
        cwd,
        dir,
        gitignore,
        absolute: true,
        dot: true,
    });
};
const firstGlob = async ({ cwd, patterns }) => {
    const stream = fg.globStream(patterns.map(removeProductionSuffix), { cwd, ignore: GLOBAL_IGNORE_PATTERNS });
    for await (const entry of stream) {
        return entry;
    }
};
const dirGlob = async ({ cwd, patterns, gitignore = true }) => globby(patterns, {
    cwd,
    dir: cwd,
    onlyDirectories: true,
    gitignore,
});
export const _glob = timerify(glob);
export const _firstGlob = timerify(firstGlob);
export const _dirGlob = timerify(dirGlob);
